### ISSUE TITLE
storage: "shared fate" does not work properly with multi-process storage clusters

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#19279]
### What version of Materialize are you using?

v0.54.0-dev

### What is the issue?

I modified Zippy to use a multi-process storaged cluster and unfortunately this does not survive all the chaos events that are being thrown at it.

In particular, the following situation occured:

```
CONTAINER ID   IMAGE                                                               COMMAND                  CREATED             STATUS                       PORTS                                                                                                                                                                                                                               NAMES
bd6f6ebef1e8   materialize/clusterd:mzbuild-AGQRZSOFCWYKZXWIJEKFCQDSZHNUH6EF       "tini -- entrypoint.…"   About an hour ago   Up 19 minutes                0.0.0.0:47506->2100/tcp, :::47495->2100/tcp, 0.0.0.0:47505->2101/tcp, :::47494->2101/tcp                                                                                                                                            zippy-storaged3-1
761b25a3f1e9   materialize/clusterd:mzbuild-AGQRZSOFCWYKZXWIJEKFCQDSZHNUH6EF       "tini -- entrypoint.…"   About an hour ago   Exited (2) 19 minutes ago                                                                                                                                                                                                                                        zippy-storaged2-1
dfa60d9aa84e   materialize/clusterd:mzbuild-AGQRZSOFCWYKZXWIJEKFCQDSZHNUH6EF       "tini -- entrypoint.…"   About an hour ago   Exited (2) 19 minutes ago                                                                                                                                                                                                                                        zippy-storaged1-1
c7f44dcb39d4   materialize/clusterd:mzbuild-AGQRZSOFCWYKZXWIJEKFCQDSZHNUH6EF       "tini -- entrypoint.…"   About an hour ago   Exited (2) 19 minutes ago                                                                                                                                                                                                                                        zippy-storaged0-1
```
In other words, 3 out of 4 storaged containers that participated in the same storage cluster are down, but one remains up. Under the "shared fate" doctrine, this should not be possible.

The storaged that remains up continues to attempt to connect to something (environmentd?):

```
2023-05-15T09:33:35.589201Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:35.623672Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:36.589698Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:36.624891Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:37.591136Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:37.627280Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:38.591871Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:38.627143Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:39.592830Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:39.628157Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:40.594593Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:40.628824Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:41.595230Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:41.630382Z  INFO mz_service::grpc: GrpcServer: remote client connected
2023-05-15T09:33:42.596669Z  INFO mz_service::grpc: GrpcServer: remote client connected
```

**Comment by aljoscha** [comment]
@guswynn might be interesting for you 🕊️

**Comment by teskje** [comment]
Well, I tried to reproduce this but couldn't. In general, the shared fate only applies once the different clusterd processes have successfully connected to each other. There is no shared fate as long as they are not connected, as nobody knows about the others' fate at that point. So I assume what happened here is that somehow the three dead clusterd processes got killed while they were not connected to the fourth one, so the fourth one survived. The connection-related logs do hit in that direction.

Closing this as not reproducible and probably fine. I don't think we've experienced issues where processes stayed alive longer than expected in production either.
